go/pkg/rexec/rexec.go

```diff
@@ -216,6 +228,271 @@ func (ec *Context) computeInputs() error {
 	return nil
 }
 
+func (ec *Context) ngUploadInputs() error {
+	cmdID, executionID := ec.cmd.Identifiers.ExecutionID, ec.cmd.Identifiers.CommandID
+	if ec.Metadata.ActionDigest.Size > 0 {
+		// Already computed inputs.
+		log.V(1).Infof("[casng] ng.req: inputs already uploaded; cmd_id=%s, exec_id=%s", cmdID, executionID)
+		return nil
+	}
+	execRoot, workingDir, remoteWorkingDir, err := cmdDirs(ec.cmd)
+	if err != nil {
+		return err
+	}
+	slo := symlinkOpts(ec.client.GrpcClient.TreeSymlinkOpts, ec.cmd.InputSpec.SymlinkBehavior)
+	filter, err := exclusionsFilter(ec.cmd.InputSpec.InputExclusions)
+	if err != nil {
+		return err
+	}
+	log.V(2).Infof("[casng] ng.req; exec_root=%s, working_dir=%s, remote_working_dir=%s, symlink_opts=%s, inputs=%d, virtual_inputs=%d, cmd_id=%s, exec_id=%s", execRoot, workingDir, remoteWorkingDir, slo, len(ec.cmd.InputSpec.Inputs), len(ec.cmd.InputSpec.VirtualInputs), cmdID, executionID)
+	log.V(4).Infof("[casng] ng.req; exec_root=%s, working_dir=%s, remote_working_dir=%s, symlink_opts=%s, inputs=%+v, virtual_inputs=%+v, cmd_id=%s, exec_id=%s", execRoot, workingDir, remoteWorkingDir, slo, ec.cmd.InputSpec.Inputs, ec.cmd.InputSpec.VirtualInputs, cmdID, executionID)
+	reqs := make([]casng.UploadRequest, 0, len(ec.cmd.InputSpec.Inputs)+len(ec.cmd.InputSpec.VirtualInputs))
+	pathSeen := make(map[impath.Absolute]bool)
+	for _, p := range ec.cmd.InputSpec.Inputs {
+		rel, err := impath.Rel(p)
+		if err != nil {
+			return err
+		}
+		absPath := execRoot.Append(rel)
+		if err != nil {
+			return err
+		}
+		// if seenPath[absPath] {
+		// return fmt.Errorf("[casng] %s %s> cannot have shared paths among inputs: %q", cmdID, executionID, absPath)
+		// }
+		pathSeen[absPath] = true
+		// Mark ancestors as seen to ensure any potential virtual parent is excluded.
+		parent := absPath.Dir()
+		for !pathSeen[parent] && parent.String() != execRoot.String() {
+			pathSeen[parent] = true
+			parent = parent.Dir()
+		}
+		reqs = append(reqs, casng.UploadRequest{Path: absPath, SymlinkOptions: slo, Exclude: filter})
+	}
+	// Append virtual inputs after real inputs in order to ignore any redundant virtual inputs.
+	// Sorting by path length descending is necessary to skip redundant ancestors. Otherwise, the conslidation in casng.UploadTree will skip descendants.
+	sort.Slice(ec.cmd.InputSpec.VirtualInputs, func(i, j int) bool {
+		return len(ec.cmd.InputSpec.VirtualInputs[i].Path) > len(ec.cmd.InputSpec.VirtualInputs[j].Path)
+	})
+	for _, p := range ec.cmd.InputSpec.VirtualInputs {
+		if p.Path == "" {
+			return fmt.Errorf("[casng] ng.req: empty virtual path; cmd_id=%s, exec_id=%s", cmdID, executionID)
+		}
+		// If execRoot is a virtual path, ignore it.
+		if p.Path == "." || p.Path == "" {
```

Comment:
This is not necessary because if `p.Path == ""` then it will return an error above